internal/ecrconsumer/image_values.go

```diff
@@ -0,0 +1,114 @@
+package ecrconsumer
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"ship-it/internal/helmrelease"
+
+	"github.com/google/go-github/github"
+)
+
+type GitCommands interface {
+	UpdateFile(msg string, branch string, path string, fileContent []byte) (*github.RepositoryContentResponse, error)
+	GetFile(branch string, path string) ([]byte, error)
+}
+
+type Image struct {
+	Registry   string
+	Repository string
+	Tag        string
+}
+
+func parseImage(repo string, tag string) (*Image, error) {
+	arr := strings.Split(repo, "/")
+
+	var registry string
+	var repository string
+	if len(arr) == 2 {
+		repository = arr[1]
+		registry = arr[0]
+	} else {
+		return nil, fmt.Errorf("repository field is invalid! (lacking either the full registry name or repository name)")
+	}
+
+	return &Image{
+		Registry:   registry,
+		Repository: repository,
+		Tag:        tag,
+	}, nil
+}
+
+func getImagePath(v reflect.Value, serviceName string) []string {
+	if v.Kind() != reflect.Map {
+		return []string{}
+	}
+
+	iter := v.MapRange()
+	for iter.Next() {
+		key := iter.Key().String()
+		if key == "image" {
+			img := v.MapIndex(iter.Key()).Interface().(map[string]interface{})
+			image, err := parseImage(img["repository"].(string), img["tag"].(string))
+			if err != nil {
+				fmt.Println("Invalid image")
+				return []string{}
+			}
+			if image.Repository == serviceName {
+				return []string{key}
+			}
+		}
+
+		val := reflect.ValueOf(iter.Value().Interface())
+		if val.Kind() == reflect.Map {
+			path := getImagePath(val, serviceName)
+			if len(path) == 0 {
+				continue
+			}
+			return append([]string{key}, path...)
+		}
+	}
+
+	return []string{}
+}
+
+func table(vals map[string]interface{}, path []string) map[string]interface{} {
+	tabled := vals
+	for i := range path {
+		_, ok := tabled[path[i]].(map[string]interface{})
+		if !ok {
+			fmt.Println("invalid path")
+			return nil
+		}
+
+		tabled = tabled[path[i]].(map[string]interface{})
+	}
+	return tabled
+}
+
+func update(vals map[string]interface{}, img Image, path []string) map[string]interface{} {
+	imgVals := table(vals, path)
+	if imgVals == nil {
+		return nil
```

Comment:
Because we return `nil` here, when the `path` is incorrect (it's passed as a param so we have no guarantees it's actually a valid path in the helmrelease's `Spec.values`), `WithImage` would return an image with empty values. 

The preferable behaviour would be to not change the vals at all if the path doesn't match, but an ideal solution would guarantee that we can never use the wrong path in the first place. As in, we wouldn't take it by a parameter to a public func, but provide it ourselves using `getImagePath`.